zebra-consensus/src/checkpoint.rs

```diff
@@ -905,9 +905,12 @@ where
         });
 
         async move {
-            commit_finalized_block
-                .await
-                .expect("commit_finalized_block should not panic")
+            let fut = commit_finalized_block.await;
+            if zebra_chain::shutdown::is_shutting_down() {
```

Comment:
I'm not sure if this shutdown check is in the right place.

`tokio::spawn` drives the task concurrently in another thread:
https://docs.rs/tokio/1.1.1/tokio/fn.spawn.html

So we can panic before the thread that runs this future gets to `commit_finalized_block.await`. I think the current code works because tokio sets `panic=unwind`. But I'm not sure if we want to rely on that.

Also, the verification in the `commit_finalized_block` future is usually much slower than this future, because it waits on other blocks. So this future usually gets to `commit_finalized_block.await` before the other future panics.

Maybe @yaahc can help?
Can we rely on tokio to always set `panic=unwind`?

Reply:
> Can we rely on tokio to always set panic=unwind?

Not in in our library crates at the very least, any dependency of ours can just use unwinding panics instead and we have no control over that. That said, I'm not following the scenario as you're laying it out. Are you talking about the panic on line 912?

Reply:
Yes, but the panic on line 912 happens because of a panic inside the spawned commit_finalized_block future, so the actual panic is probably on line 897. (Or it's a tokio task cancellation due to shutdown.)

Reply:
At the moment, I think that there's a panic inside commit_finalized_block that gets turned into JoinError::Cancelled by tokio, but I'm not sure.

If there is a panic inside commit_finalized_block, and it could ever turn into an abort, we need to check the shutdown flag there.

Reply:
I don't think there are any soundness issues here regarding the panic personality. With panic = abort we can see the spawned thread panic and immediately take down the entire application. With panic = unwind we will instead capture the panic, and re-panic in the else case on line 912. As far as the API of `zebra-consensus` is concerned this is fine. The main issue is that when you're capturing panics and re-throwing them it tends to be lossy and slow, so depending on how a consumer of `zebra-consensus` uses the `Service` they will need to handle the case where the service goes down, but that should be possible with our current API.

Reply:
I'm not concerned about soundness or even state correctness issues here.

We check the shutdown flag here to avoid a panic on Ctrl-C. But this is the wrong place to check the shutdown flag if we're actually panicking somewhere else.

Specifically, we might need to check the flag elsewhere if:
1. we `panic=abort` on either `.expect("Verified checkpoints must be committed transactionally")`
2. whatever is causing the `JoinError::Cancelled` turns into a panic

I think this is a complex enough issue that it at least deserves a clarifying comment. I don't think we need to make any further code changes right now, because the code seems to work, and this is a low-priority fix.

Reply:
Here's a draft comment.
```suggestion
            // Avoid a panic on shutdown
            //
            // When `zebrad` is terminated using Ctrl-C, the `commit_finalized_block` task
            // can return a `JoinError::Cancelled`. We expect task cancellation on shutdown,
            // so we don't need to panic here. The persistent state is correct even when the
            // task is cancelled, because block data is committed inside transactions, in
            // height order.
            if zebra_chain::shutdown::is_shutting_down() {
```

Reply:
@yaahc #1576 is the bug we're trying to fix here

Reply:
aah, that makes sense. We might need to put the check for `is_shutting_down` in the panic hook itself.

Reply:
https://github.com/oxarbitrage/zebra/blob/issue1574/zebrad/src/application.rs#L222-L224

This would be the place, where we just skip reporting any panics if the shutdown flag is enabled.

Reply:
Hmm yeah I hadn't thought of that. Good idea!

I don't think we want to skip all panics on shutdown, but it would definitely be easier to filter out certain kinds of panics (for example: `JoinError::Cancelled`) in the panic hook.

Can you open a ticket for that?

This PR fixes the known issues right now, and I don't think any further changes are urgent. We can move to the panic hook next time we find a shutdown panic.

Reply:
For example, if a panic indicates a corrupted state, we don't want to skip it at shutdown.

Reply:
(I've just committed the comment, because GitHub is hiding the middle of this conversation.)

Reply:
> I don't think we want to skip all panics on shutdown, but it would definitely be easier to filter out certain kinds of panics (for example: JoinError::Cancelled) in the panic hook.

The main issue is that panics completely erase the type of the payload, usually by stringifying it, so filtering is gonna be string matching :/. My feeling is that we're not planning on keeping this code long term anyways, and if there's corruption in the state I would expect the error to be triggered on the next start, not while shutting down. It doesn't seem like a huge risk to me, versus the benefit of pretty consistently suppressing superfluous errors until we've got graceful shutdown implemented.

Reply:
I think this discussion belongs in a new ticket :-)

Reply:
This refactor is now #1677